src/gc.c

```diff
@@ -3499,6 +3515,27 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
         else {
             ptls2->heap.remset->len = 0;
         }
+        // free empty GC state for threads that have exited
+        if (jl_atomic_load_relaxed(&ptls2->current_task) == NULL) {
```

Comment:
We probably need to check that the thread for which we're doing the cleanup is actually a mutator thread (i.e. `ptls2->tid < gc_first_tid  || ptls2->tid >= gc_first_tid + jl_n_gcthreads`).

Note that, since GC threads have no tasks associated with them, this version is also doing the cleanup for the GC threads, and will set `gc_all_tls_states[<SOME_GC_THREAD_TID>]` to `NULL`, causing us to "lose" that GC thread.